traffic_ops/testing/api_contract/v4/test_cachegroups.py

```diff
@@ -0,0 +1,100 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""API Contract Test Case for cachegroup endpoint."""
+import logging
+import pytest
+import requests
+
+from trafficops.tosession import TOSession
+
+# Create and configure logger
+logger = logging.getLogger()
+
+primitive = bool | int | float | str | None
+
+@pytest.mark.parametrize('request_template_data', ["cachegroup"], indirect=True)
+def test_cachegroup_contract(to_session: TOSession, request_template_data:
+	list[dict[str, object] | list[object] | primitive],
+	response_template_data: dict[str, object],cachegroup_post_data: dict[str, object]) -> None:
+	"""
+	Test step to validate keys, values and data types from cachegroup endpoint
+	response.
+	:param to_session: Fixture to get Traffic Ops session.
+	:param request_template_data: Fixture to get request template data from a prerequisites file.
+	:param response_template_data: Fixture to get response template data from a prerequisites file.
+	:param cachegroup_post_data: Fixture to get sample cachegroup data and actual cachegroup response.
+	"""
+	# validate CDN keys from cdns get response
+	logger.info("Accessing /cachegroup endpoint through Traffic ops session.")
+
+	cachegroup = request_template_data[0]
+	if not isinstance(cachegroup, dict):
+		raise TypeError("malformed cachegroup in prerequisite data; not an object")
+
+	cachegroup_name = cachegroup.get("name")
+	if not isinstance(cachegroup_name, str):
+		raise TypeError("malformed cachegroup in prerequisite data; 'name' not a string")
+
+	cachegroup_get_response: tuple[
+		dict[str, object] | list[dict[str, object] | list[object] | primitive] | primitive,
+		requests.Response
+	] = to_session.get_cachegroups(query_params={"name": str(cachegroup_name)})
+	try:
+		cachegroup_data = cachegroup_get_response[0]
+		if not isinstance(cachegroup_data, list):
+			raise TypeError("malformed API response; 'response' property not an array")
+
+		first_cachegroup = cachegroup_data[0]
+		if not isinstance(first_cachegroup, dict):
+			raise TypeError("malformed API response; first Cache group in response is not an object")
+		cachegroup_keys = set(first_cachegroup.keys())
+
+		logger.info("Cache group Keys from cachegroup endpoint response %s", cachegroup_keys)
+		response_template = response_template_data.get("cachegroup").get("properties")
```

Comment:
> Sorry, My mistake, The type of response_template_data should be dict[str, object], Updated this in code.

Actually, you're doing a `.get` on the result of that `.get`. Then you store that result and later do a `.get` on it and then a `.get` on *that*. So the type has to be `dict[str, dict[str, dict[str, dict[str, object]]]]` or something more specific assignable to that.

But updating that typing isn't enough to completely fix it, because since you're using `.get` to do safe access, you can get `None` in addition to the value type. So right now you have `response_template_data` which has the type `dict[str, dict[str, dict[str, dict[str, object]]]]`, and therefore `response_template_data.get("cachegroup")` has the type `dict[str, dict[str, dict[str, object]]] | None`, and you can't do `.get` on that type because `None` has no `get` method. You need to deal with the possibility that the data you're accessing doesn't exist. You can do that by checking for `None` a bunch, or you can do it by `except`-ing a `KeyError` (or even not `except`-ing it if the caller can be expected to handle it well enough).

Reply:
Added checks for the response template data. added except -ing a Keyerror.